### ISSUE TITLE
Centralized asset auto-export/auto-dumping for facilitating sync between team members

### ISSUE THREAD

**Issue by Obsdark** [stride3d/stride#1634]
**Is your feature request related to a problem? Please describe.**
Is a suggestion to improve merging between versions of the projects from different team members, and help to reduce, detect and fix very easily and fast any sync problems between such versions.

**Describe the solution you'd like**
Have some kind of natural text file somewhere in the file than save on itself in somewhat of a readable languaje, all the scenes, prefabs, models and such you can find in the project.

Why is that?

Because most of the assets in stride are already text files in readable language than reference resources somewhere in the machine and work them somehow, i'm talking about, for instance, SpriteSheets, Models, prefabs, you name it, but this assets are distributed around the editor folders in different sub-folders and such, so in order to make projects easier to merge between different team members, this suggestion is to have a file than can be merged through text, (like using git or maybe using WinMerge or something similar) than can work as a save-state of the complete project in that regard, in order to make the merge of this file the only necessary thing to have two identical project from the assets perspective, (besides actually having the referenced files for the SpriteSheets or similar of course)

Think something similar to the GameSettings file, which is there in the folder of the project being called GameSettings.sdgamesettings basically floating in the space, and that's perfectly fine and kinda the point.

Apparently stride already have something like that but saved internally, so the suggestion will be to convert that in a file like GameSettings, or make somewhat of a dump of that db every time there are changes in the assets and also detect and proposefully load it every time the file is changed externally. 

Readability is only necessary for this suggestion up to the point to make possible manual merging (with WinForms for instance), it doesn't need to be really readable to the fullest if that were an issue in implementation.

That way a team can have easy merging, being the case the guys are using git, or being the case the artists are using winrar and sharing the project by hand, anyone could be able to merge the projects in a very flexible and reliable manner with this change and i would say could make far, far easier to work for teams. 

**Additional context**
I would think than improve the experience of working as a team in a better coordinated way and with better sync without spending in cloud for instance is like one of the things than Stride could do than other engines can't, because at least in some relevant cases, doing so will go against his business model (of the other engines).

In general we continue talk about how we will like than Stride could be a more well known engine, and i think that conversation is fine, but there is up to this moment no real difference or reason to really use Stride over other similar C# engines besides security, chance to modify things yourself and somewhat efficiency, which on itself aren't things in the general public eye, or even industry eye than are consider as an advantage all by themselves, industry for instance can have that modifications chance and support paying the premium of some of this other engines for instance to get that same services.

Something like this could make it easier to coordinate between team members, sync, and because of that make the times of development faster, which in time can make it a better candidate for both, gamejams and teams than will want to make something to drop on the market giving Stride3d more chances to be more used, and become better known.

**Comment by manio143** [comment]
I'm not clear how having a file with a list of assets in the project improves merging projects over having asset files in a directory (without an explicit index).

At the moment to merge two projects you just merge their assets folders and if there's no conflicts (i.e. there's no two same paths in those projects) that's it.

This is already nicely supported by git, which could furthermore highlight merge conflicts.

**Comment by Obsdark** [comment]
I understand your point, and indeed for cases where git is required or a custom as is for most people using stride already the problem can be solved with git, however that is not the only kind of people who use game engines and the idea is precisely to expand the use of it.

Think in the case of people who share files through winrar for instance, artist which do not use git and/or a team who work in a gamejam, being such the case you can inmediatly see the problems with it, at least i remember clearly last time i tried some versions ago, some of them were:

- If you move the assets from one folder to the other (drag & drop) is not always detected.

- If two or more members of a team doesn't have access to the same folders, or if you make something isolated before join it to the rest of the job this can be an issue, specially if the items are in different folders, because in that case you will need to manually select the right item and manually order it and merge it, this can, by the lowest point, leave some files around being garbage, for the highest, make you use redundant materials or even the wrong material with the same name for the prefab or so you want to use it.

I use the example of materials but you get the jist, and i will asume we want to help people who work "less formally", that people is a big group and they tend to use other certain c# engine than i'm not gonna name directly here...

**Comment by manio143** [comment]
Currently all assets have Guid identifiers, so moving files across folders shouldn't break references between assets (if it does that's a separate bug).

For resource files we currently don't track them in any way, so if you replace a PNG texture with another, you wouldn't be able to tell (although for some assets I think we may be storing a hash of the referenced resource to detect changes, but not in a breaking your build kind of way).

Am I understanding you correctly that you're advocating for a single file describing all assets in a project instead of each asset being in a separate file?
I believe that would have a detrimental effect on the managing of projects rather than positive and would make it harder to implement some performance boosting features for the editor.

The problem of merging projects should ideally be solved by writing better tooling around the existing asset file system unless there's problems with the design of the system in the first place. I don't think user experience should drive implementation details. And it sounds to me like this is the case: a user unfamiliar with the design of directory based indexing would have a hard time performing an action. Creating a small executable program which performs project merging or at least validates if projects have been merged correctly would be a better way forward here.

If we're talking about creating an export/import feature in the editor/cli - we can go about designing it in whatever way we see fit - the file format should not be editable by hand and the program takes care of ensuring project coherence on import.
If we're talking about changing project structure in general, I'm opposed to that because I haven't yet seen an explanation how it would be better than the current system wrt the tooling most devs use. Stride is inherently a member of the .NET eco system where using git is prevalent. Rather than retrofitting the engine into needs of less tech savvy, we should enable them by providing user friendly features (i.e. not manual merging of projects).

**Comment by Obsdark** [comment]
Agreed in several points if not all, except than -even if it's confusing- i'm not specifically advocating for a single file describing all assets, i do propose that, yes, in my reply just before, but i think it would be equally beneficial if instead we have as you said some kind of dump file than allow for easy restoration and merging of two projects between themselfs, that's why the title of the suggestion is as it is.

Agreed also with your point of, as VS actually does, add some friendly features to enable the less tech savy into being able to solve that issue in a more correct manner, and as such another thing than could be maybie more tiresome of program but could be the solution to that is allow for some kind of configuration file than allow asset placing enforcing.

What i mean by that is than, you can, using an UI or otherwise, tell Stride which folder inside the asset folder is for what type of asset, therefore making than when habilitated, doesn't matter where you drag and drop that material because it will be moved to the correct folder anyway and it will be set behind the curtain with the correct prefab/model/wherever if apply to the case.

Also can be done than this "asset placing enforcing" config file can be checked if it doesn't exist, and if somebody add it, then, it habilitate his functionality in the config file and, if drag & drop another version of it when one older existe, it replace it with the new one, it could be even able to be sended through git too, been a config file and all of that.

That could make any manual merging far far easier, extra points if the UI gets an option to merge two projects like if it was the load or new button in the gameeditor, which could use this asset placing enforcing config files like the mark to define how the assets should be ordered after the fusion of the project, and how they should be connected maybie.

This is of course a third single file ordering suggestion, which is kinda my point, not the single file part by itself but some way to make the use of the new feature really really easy in the hands of the less tech savy people, i mean, i can see in a gamejam somebody telling another person to "drag & drop this file to the gameeditor and accept everything", that is easy enough, or the programmer moving that file manually to have more clarity about something, or simply don't do any drag and drop because when using the UI in the "merge" option, the developer use this file in order to force the order of the new project than the artist send him and been able to use it confortly.
